msgraph-proxy-plugins/MocksResponses/MockResponsePlugin.cs

```diff
@@ -0,0 +1,161 @@
+using System.CommandLine.Invocation;
+using System.Net;
+using System.Text.Json;
+using System.Text.Json.Serialization;
+using System.Text.RegularExpressions;
+using Titanium.Web.Proxy.EventArguments;
+using Titanium.Web.Proxy.Http;
+using Titanium.Web.Proxy.Models;
+
+namespace Microsoft.Graph.DeveloperProxy.Plugins.MocksResponses;
+
+internal class MockResponseConfiguration {
+    public bool NoMocks { get; set; } = false;
+    public string MocksFile { get; set; } = "responses.json";
+
+    [JsonPropertyName("responses")]
+    public IEnumerable<MockResponse> Responses { get; set; } = Array.Empty<MockResponse>();
+}
+
+public class MockResponsePlugin : IProxyPlugin {
+    private ISet<Regex>? _urlsToWatch;
+    private ILogger? _logger;
+    private MockResponseConfiguration _configuration = new();
+    private MockResponsesLoader? _loader = null;
+    private readonly Option<bool> _noMocks;
+    public string Name => nameof(MockResponsePlugin);
+
+    public MockResponsePlugin() {
+        _noMocks = new Option<bool>("--no-mocks", "Disable loading mock requests");
+        _noMocks.AddAlias("-n");
+        _noMocks.ArgumentHelpName = "no mocks";
+        _noMocks.SetDefaultValue(false);
+    }
+
+    public void Register(IPluginEvents pluginEvents,
+                            IProxyContext context,
+                            ISet<Regex> urlsToWatch,
+                            IConfigurationSection? configSection = null) {
+        if (pluginEvents is null) {
+            throw new ArgumentNullException(nameof(pluginEvents));
+        }
+
+        if (context is null || context.Logger is null) {
+            throw new ArgumentException($"{nameof(context)} must not be null and must supply a non-null Logger", nameof(context));
+        }
+
+        if (urlsToWatch is null || urlsToWatch.Count == 0) {
+            throw new ArgumentException($"{nameof(urlsToWatch)} cannot be null or empty", nameof(urlsToWatch));
+        }
+
+        _urlsToWatch = urlsToWatch;
+        _logger = context.Logger;
+        configSection?.Bind(_configuration);
+        _loader = new MockResponsesLoader(_logger, _configuration);
+        _loader.InitResponsesWatcher();
+
+        pluginEvents.Init += OnInit;
+        pluginEvents.OptionsLoaded += OnOptionsLoaded;
+        pluginEvents.Request += OnRequest;
+    }
+
+    private void OnInit(object? sender, InitArgs e) {
+        e.RootCommand.AddOption(_noMocks);
+    }
+
+    private void OnOptionsLoaded(object? sender, OptionsLoadedArgs e) {
+        InvocationContext context = e.Context;
+        // allow disabling of mocks as a command line option
+        _configuration.NoMocks = context.ParseResult.GetValueForOption(_noMocks);
+    }
+
+    private void OnRequest(object? sender, ProxyRequestArgs e) {
+        Request request = e.Session.HttpClient.Request;
+        ResponseState state = e.ResponseState;
+        if (!_configuration.NoMocks && _urlsToWatch is not null && e.ShouldExecute(_urlsToWatch)) {
```

Comment:
Shouldn't we also check here if the response has already been set to avoid overwriting another plugin?